### PR TITLE
Fix #138 Add medium level of IntelliSense / limit to the analysis depth (Related to #88).

### LINKED ISSUE #138: Medium-power IntelliSense mode
This would be helpful for users who are more CPU-bound, but still want some degree of IntelliSense. Right now the only options are on/off.

### PR DESCRIPTION
**Pull request by kant2002**
Limit depth analysys to relative paths which contains only 2 node_modules. This should be sufficient enough to capture
a) Capture API of the root package
b) Capture API of the dependencies. This is obiviously the must to analyze (1 node_modules)
c) Capture API of subdependencies (2 node_modules). This is also the must, since a lot of packages are wrapper/glue for another library. For example using gulp-typescript require that gulp-typescript should be parsed (level b) and typescript should be parsed (level c) since typescritp provide some config options for gulp-typescript.

Related to #88 

#138 Added Medium level of Intellisense. This level now same as Full level, but limit analysis depth to the 2 nested modules depth. Full mode is now has limit to 4 modules depth which I almost sure practical enough, but could be increased if needed. Low level is analyse 1 level depth.

### CODE REVIEW

**Comment by kant2002** [comment]
I add the configuration option for the Middle level of Intellisense (#138). Right now analysis level not configurable, but if you think this would be valuable, I could put additional config option for that.

**Comment by mousetraps** [comment]
I'll take a look at this PR when I've had more sleep :smiley:. But lets make this configurable in the per-project setting as well - it shouldn't override the users general setting, but rather serve as an additional filter. This would enable users who are sharing a project to figure out good project-specific settings rather than relying on every single person on the team to use the same defaults. Because ultimately this is a project-specific perf issue. 

For the sake of UI simplicity, Let's hold off on making it configurable from general project settings unless/until we start getting requests for it.

File: Nodejs/Product/Nodejs/Options/NodejsIntellisenseOptionsControl.resx [hunk 1]
```diff
@@ -322,6 +322,51 @@
...
+    <value>8, 33</value>
+  </data>
+  <data name="_mediumIntelliSenseRadioButton.Margin" type="System.Windows.Forms.Padding, System.Windows.Forms">
+    <value>8, 4, 8, 4</value>
+  </data>
+  <data name="_mediumIntelliSenseRadioButton.Size" type="System.Drawing.Size, System.Drawing">
+    <value>154, 21</value>
+  </data>
+  <data name="_mediumIntelliSenseRadioButton.TabIndex" type="System.Int32, mscorlib">
+    <value>5</value>
```
Comment (mousetraps) [inline Nodejs/Product/Nodejs/Options/NodejsIntellisenseOptionsControl.resx]: Check the tab index here. WinForms is annoying in that it won't automatically update all the controls on the page.

File: Nodejs/Tests/AnalysisTests/AnalysisFile.cs [hunk 2]
```diff
@@ -131,6 +131,21 @@ static class Analysis {
             List<AnalysisFile> files = new List<AnalysisFile>();
             foreach (var file in Directory.GetFiles(directory, "*", SearchOption.AllDirectories)) {
                 if (String.Equals(Path.GetExtension(file), ".js", StringComparison.OrdinalIgnoreCase)) {
+                    var relativeFile = file.Substring(directory.Length);
+                    int nestedModulesCount = 0;
+                    int index = relativeFile.IndexOf("node_modules");
+                    while (index != -1)
+                    {
```
Comment (mousetraps) [inline Nodejs/Tests/AnalysisTests/AnalysisFile.cs]: brace on same line here and everywhere else

File: Nodejs/Product/Nodejs/Options/NodejsIntellisenseOptionsControl.cs [hunk 3]
```diff
@@ -43,7 +43,11 @@ public partial class NodejsIntellisenseOptionsControl : UserControl {
                 if (_fullIntelliSenseRadioButton.Checked) {
                     return AnalysisLevel.High;
                 } else {
-                    return AnalysisLevel.None;
+                    if (_mediumIntelliSenseRadioButton.Checked) {
```
Comment (mousetraps) [inline Nodejs/Product/Nodejs/Options/NodejsIntellisenseOptionsControl.cs]: this if/else block doesn't need to be nested

File: Nodejs/Product/Analysis/Analysis/AnalysisLimits.cs [hunk 4]
```diff
@@ -34,6 +34,17 @@ internal class AnalysisLimits {
             MaxObjectLiteralProperties = 50;
             MaxObjectKeysTypes = 5;
             MaxMergeTypes = 5;
+            NestedModulesLimit = 4;
```
Comment (mousetraps) [inline Nodejs/Product/Analysis/Analysis/AnalysisLimits.cs]: Do we really want to mess with the high analysis limits here? How did you decide on 4?

File: Nodejs/Tests/AnalysisTests/AnalysisFile.cs [hunk 5]
```diff
@@ -131,6 +131,21 @@ static class Analysis {
             List<AnalysisFile> files = new List<AnalysisFile>();
             foreach (var file in Directory.GetFiles(directory, "*", SearchOption.AllDirectories)) {
                 if (String.Equals(Path.GetExtension(file), ".js", StringComparison.OrdinalIgnoreCase)) {
+                    var relativeFile = file.Substring(directory.Length);
+                    int nestedModulesCount = 0;
+                    int index = relativeFile.IndexOf("node_modules");
```
Comment (mousetraps) [inline Nodejs/Tests/AnalysisTests/AnalysisFile.cs]: use StringComparison.OrdinalIgnoreCase for this. Also we have this in a resource string, so may as well use that (NodejsConstants.NodeModulesFolder)

File: Nodejs/Tests/AnalysisTests/AnalysisFile.cs [hunk 6]
```diff
@@ -131,6 +131,21 @@ static class Analysis {
...
+                    while (index != -1)
+                    {
+                        nestedModulesCount++;
+                        relativeFile = relativeFile.Substring(index + 1);
```
Comment (mousetraps) [inline Nodejs/Tests/AnalysisTests/AnalysisFile.cs]: instead of creating a whole new string, you can change the search position in IndexOf

File: Nodejs/Tests/AnalysisTests/AnalysisFile.cs [hunk 7]
```diff
@@ -131,6 +131,21 @@ static class Analysis {
...
+                        index = relativeFile.IndexOf("node_modules");
+                    }
+
+                    if (nestedModulesCount > limits.NestedModulesLimit)
```
Comment (mousetraps) [inline Nodejs/Tests/AnalysisTests/AnalysisFile.cs]: You should be able figure this out before you've traversed the entire string.

File: Nodejs/Product/Analysis/Analysis/AnalysisLimits.cs (hunk 4 above)
Comment (kant2002) [inline Nodejs/Product/Analysis/Analysis/AnalysisLimits.cs]: Reason for 4 following - I try to be very practical about numbers. 

For small applications this number more then enough, and they don't have any issues dealing with some potential errors in Intellisense.

But if we have large application with Full setting, then VS Tools for Node.JS as a product immediately became unusable for that users.

I try to imagine dependency tree which exceed this 4 depth limit. Here the examples of trees about which I thought:
## Example 1
Level 0 - Large system. Some code should use properties of the object on level 4 here. 
Level 1 - Subsystem - pass data from level 4
Level 2 - Internal dependency for company - Do some work and pass data to level 1
Level 3 - Framework on top of which created Internal dependency.
Level 4 - Dependency of the framework. Objects from here still should be available.
Level 5 - Dependency of the framework provide some usefull primitive which would be used very often on the level of whole system. Hmmm. 

## Example 2 (reason why I could increase to 5)
Level 0 - Large system. Some code should use properties of the object on level 4 here. 
Level 1 - Subsystem - pass data from level 4
Level 2 - Internal dependency for company - Wrap access to internal library and perform business logic. Do some work and pass data to level 1
Level 3 - Internal library which wrap access to API.
Level 4 - Http library.
Level 5 - Promise Polyfill. 

All these examples are highly speculative and I specifically try to create such deep level. If you develop on windows with such deep level you already close to your limit, your maximum is probably 10. 
If you have such deep structure already, chances that you have more then one such deep dependency is very high, and thus we just increase time for loading solution, loading analysis file and increase memory consumption without any practical benefits.

From my point of view practical benefits much better, and better have good and usable product with 90% working intellisense.
Or we just have 2 options: Medium and None without Full being usable by anyone on the large projects.

File: Nodejs/Product/Nodejs/Options/NodejsIntellisenseOptionsControl.cs (hunk 3 above)
Comment (kant2002) [inline Nodejs/Product/Nodejs/Options/NodejsIntellisenseOptionsControl.cs]: Should I create something like this?

```
if (_fullIntelliSenseRadioButton.Checked) {
    return AnalysisLevel.High;
}

if (_mediumIntelliSenseRadioButton.Checked) {
    return AnalysisLevel.Medium;
} else {
    return AnalysisLevel.None;
}
```

File: Nodejs/Product/Nodejs/Options/NodejsIntellisenseOptionsControl.resx [hunk 8]
```diff
@@ -322,6 +322,51 @@
...
+  </data>
+  <data name="_mediumIntelliSenseRadioButton.Text" xml:space="preserve">
+    <value>&amp;Medium IntelliSense</value>
+  </data>
+  <data name="_mediumIntelliSenseRadioButton.TextAlign" type="System.Drawing.ContentAlignment, System.Drawing">
+    <value>TopLeft</value>
+  </data>
+  <data name="_mediumIntelliSenseRadioButton.ToolTip" xml:space="preserve">
+    <value>Medium completions (less memory/CPU usage)</value>
+  </data>
```
Comment (mousetraps) [inline Nodejs/Product/Nodejs/Options/NodejsIntellisenseOptionsControl.resx]: Minor language suggestions, since I'm just now realizing that our label for "high" isn't actually "high", and we don't have a "low", so "medium" doesn't work as well. 
* Name: Quick IntelliSense
* Tooltip: Limited completions (reduced memory/CPU usage)

File: Nodejs/Product/Nodejs/Options/NodejsIntellisenseOptionsControl.cs (hunk 3 above)
Comment (mousetraps) [inline Nodejs/Product/Nodejs/Options/NodejsIntellisenseOptionsControl.cs]: Yeah, and you can even go one step further with if, else if, else for best readability.

**Comment by mousetraps** [comment]
Left some comments, minor tweaks for the most part. @DinoV, it would be great to get your eyes on this as well.

File: Nodejs/Product/Analysis/Analysis/AnalysisLimits.cs (hunk 4 above)
Comment (mousetraps) [inline Nodejs/Product/Analysis/Analysis/AnalysisLimits.cs]: Awesome, thanks for the detailed explanation. Logically it makes sense to me, though Dino has the  most familiarity on this and what the edge cases might be, so it'll be great if he can chime in here too.

File: Nodejs/Tests/AnalysisTests/AnalysisFile.cs (hunk 7 above)
Comment (kant2002) [inline Nodejs/Tests/AnalysisTests/AnalysisFile.cs]: If I push another commit with changes, would our discussion would be preserved?

Comment (mousetraps) [inline Nodejs/Tests/AnalysisTests/AnalysisFile.cs]: Yep, and that's preferred so we can have the full discussion in one place. you won't see it inline with the current code, but the old diff will still be available by each comment.

File: Nodejs/Tests/AnalysisTests/AnalysisFile.cs [hunk 9]
```diff
@@ -131,6 +131,20 @@ static class Analysis {
...
             foreach (var file in Directory.GetFiles(directory, "*", SearchOption.AllDirectories)) {
                 if (String.Equals(Path.GetExtension(file), ".js", StringComparison.OrdinalIgnoreCase)) {
+                    var relativeFile = file.Substring(directory.Length);
+                    int nestedModulesCount = 0;
+                    int startIndex = 0;
+                    int index = relativeFile.IndexOf("node_modules", startIndex, StringComparison.OrdinalIgnoreCase);
+                    while (index != -1) {
+                        nestedModulesCount++;
+                        if (nestedModulesCount > limits.NestedModulesLimit) {
+                            continue;
```
Comment (mousetraps) [inline Nodejs/Tests/AnalysisTests/AnalysisFile.cs]: the check is nested now, so continue is going to continue the while loop, not the outer for loop, so the while loop just became infinite after the limit is exceeded. 

One solution is the refactor the while loop out into its own method, say "NestedModulesExceedsLimit", and return true instead of continuing, false otherwise. Then...
```csharp
if (!NestedModulesExceedsLimit(...) { 
    files.Add(new AnalysisFile(file, File.ReadAllText(file)));
}
```

File: Nodejs/Tests/AnalysisTests/AnalysisFile.cs [hunk 10]
```diff
@@ -131,6 +131,20 @@ static class Analysis {
             List<AnalysisFile> files = new List<AnalysisFile>();
             foreach (var file in Directory.GetFiles(directory, "*", SearchOption.AllDirectories)) {
                 if (String.Equals(Path.GetExtension(file), ".js", StringComparison.OrdinalIgnoreCase)) {
+                    var relativeFile = file.Substring(directory.Length);
```
Comment (mousetraps) [inline Nodejs/Tests/AnalysisTests/AnalysisFile.cs]: Because you are using the position of the string, you can get rid of this substring completely, and just use file instead of relativeFile.

File: Nodejs/Tests/AnalysisTests/AnalysisFile.cs [hunk 11]
```diff
@@ -131,6 +131,20 @@ static class Analysis {
...
+                    int nestedModulesCount = 0;
+                    int startIndex = 0;
+                    int index = relativeFile.IndexOf("node_modules", startIndex, StringComparison.OrdinalIgnoreCase);
+                    while (index != -1) {
+                        nestedModulesCount++;
+                        if (nestedModulesCount > limits.NestedModulesLimit) {
+                            continue;
+                        }
+
+                        startIndex = index + 1;
```
Comment (mousetraps) [inline Nodejs/Tests/AnalysisTests/AnalysisFile.cs]: this can be ```index + "node_modules".Length```

File: Nodejs/Product/Analysis/Analysis/AnalysisLimits.cs (hunk 4 above)
Comment (DinoV) [inline Nodejs/Product/Analysis/Analysis/AnalysisLimits.cs]: I'm not seeing the change that actually limits this in the product?  AnalysisTests\AnalysisFile will limit it when we run the tests, but won't do anything when running inside of VS.

Personally I would suggest not putting this in AnalysisLimits, and not having the analyzer actually be aware of this at all.  The AnalysisLimits class is really used for limits on things when we're analyzing.  But in this case we actually just don't want to analyze these files at all.  Which means we also want to avoid parsing them entirely too.  

Therefore I think the correct spot to put this limit in would be in NodejsFileNode.ShouldAnalyze.  If you return false from there then we won't even bother parsing the file which should give the best performance gain.

**Comment by mousetraps** [comment]
Hey there, just checking in to see if you have any questions. It would be great to get this change into the next release. :smiley:

**Comment by kant2002** [comment]
Hi, was side-tracked on the main work. Will take a look at the comments from @DinoV later today, or tomorrow

**Comment by mousetraps** [comment]
Yeah, no worries - it's totally okay to have a life outside NTVS :wink:. This feature is super impactful, though, so unlike cookie dough... the more bake time, the better.

**Comment by objelke** [comment]
>> This feature is super impactful, though

oh yes, we have a good amount of developers waiting... (the homebrew patch is fine but it would be so much nicer to get rid of it)
If I can assist, most happy to do so. Let me know.

**Comment by kant2002** [comment]
Thanks all for waiting, I send additional changes which should address @DinoV comment that I don't place fix in the correct place.

File: Nodejs/Product/Nodejs/Project/NodejsProjectNode.cs [hunk 12]
```diff
@@ -490,6 +490,26 @@ public NodejsProjectNode(NodejsProjectPackage package)
...
+            int nestedModulesCount = 0;
+            int startIndex = 0;
+            int index = relativeFile.IndexOf(AnalysisConstants.NodeModulesFolder, startIndex, StringComparison.OrdinalIgnoreCase);
+            while (index != -1) {
+                nestedModulesCount++;
+                if (nestedModulesCount > maxModulesDepth){
+                    return false;
+                }
+
+                startIndex = index + 1;
```
Comment (mousetraps) [inline Nodejs/Product/Nodejs/Project/NodejsProjectNode.cs]: this can be startIndex = ```index + "node_modules".Length``` since you already know you've seen node_modules before.

File: Nodejs/Tests/AnalysisTests/AnalysisFile.cs [hunk 13]
```diff
@@ -131,6 +131,20 @@ static class Analysis {
             List<AnalysisFile> files = new List<AnalysisFile>();
             foreach (var file in Directory.GetFiles(directory, "*", SearchOption.AllDirectories)) {
```
Comment (mousetraps) [inline Nodejs/Tests/AnalysisTests/AnalysisFile.cs]: You can remove the code from here now, right?

File: Nodejs/Product/Analysis/Analysis/AnalysisLimits.cs [hunk 14]
```diff
@@ -34,6 +34,17 @@ internal class AnalysisLimits {
             MaxObjectLiteralProperties = 50;
             MaxObjectKeysTypes = 5;
             MaxMergeTypes = 5;
+            NestedModulesLimit = AnalysisConstants.MaxAnalysisDepthQualty;
```
Comment (mousetraps) [inline Nodejs/Product/Analysis/Analysis/AnalysisLimits.cs]: For the sake of readability and consistency, let's not add this to AnalysisConstants, and instead just include the number. The comment about reasoning is very helpful, though, and should be included where the NestedModulesLimit property is defined.

File: Nodejs/Product/Nodejs/Project/NodejsProjectNode.cs [hunk 15]
```diff
@@ -490,6 +490,26 @@ public NodejsProjectNode(NodejsProjectPackage package)
             if (CommonUtils.IsSubpathOf(_intermediateOutputPath, fileNode.Url)) {
                 return false;
             }
+
+            int maxModulesDepth = AnalysisConstants.MaxAnalysisDepthQualty;
```
Comment (mousetraps) [inline Nodejs/Product/Nodejs/Project/NodejsProjectNode.cs]: Better yet, use ```Analyzer.Project.Limits.NestedModulesLimit``` to get the current limit.

File: Nodejs/Product/Nodejs/Project/NodejsProjectNode.cs [hunk 16]
```diff
@@ -490,6 +490,26 @@ public NodejsProjectNode(NodejsProjectPackage package)
             if (CommonUtils.IsSubpathOf(_intermediateOutputPath, fileNode.Url)) {
                 return false;
             }
+
```
Comment (mousetraps) [inline Nodejs/Product/Nodejs/Project/NodejsProjectNode.cs]: This entire block should be placed below the checks for _maxFileSize and _analysisIgnoredDirs, not above.

File: Nodejs/Product/Nodejs/Project/NodejsProjectNode.cs [hunk 17]
```diff
@@ -490,6 +490,26 @@ public NodejsProjectNode(NodejsProjectPackage package)
...
+            while (index != -1) {
+                nestedModulesCount++;
+                if (nestedModulesCount > maxModulesDepth){
+                    return false;
+                }
+
+                startIndex = index + 1;
+                index = relativeFile.IndexOf(AnalysisConstants.NodeModulesFolder, startIndex, StringComparison.OrdinalIgnoreCase);
+            }
+
```
Comment (mousetraps) [inline Nodejs/Product/Nodejs/Project/NodejsProjectNode.cs]: This entire block should be placed below the checks for _analysisIgnoredDirs and _maxFileSize, rather than above because those should take precedent.

**Comment by mousetraps** [comment]
Awesome, thanks! Added some more comments, so lgtm after that. DinoV and @objelke, feel free to jump in if you have any more comments.

File: Nodejs/Tests/AnalysisTests/AnalysisFile.cs (hunk 13 above)
Comment (kant2002) [inline Nodejs/Tests/AnalysisTests/AnalysisFile.cs]: I think this code should still be there for testing reason. 
Without that code we could measure just raw JsAnalyzer performance without any limitations, but there should be way to measure analysis speed with depth limit. If needed I could add command line switch to be able distinguish between desired behaviors.

File: Nodejs/Product/Nodejs/Project/NodejsProjectNode.cs (hunk 15 above)
Comment (kant2002) [inline Nodejs/Product/Nodejs/Project/NodejsProjectNode.cs]: Thanks!! I was looking for something like that.

File: Nodejs/Tests/AnalysisTests/AnalysisFile.cs (hunk 13 above)
Comment (mousetraps) [inline Nodejs/Tests/AnalysisTests/AnalysisFile.cs]: How about a new unit test?

Comment (mousetraps) [inline Nodejs/Tests/AnalysisTests/AnalysisFile.cs]: The issue with the test as its written now is that it's not actually testing the code that's in the product itself.

Comment (kant2002) [inline Nodejs/Tests/AnalysisTests/AnalysisFile.cs]: From what I understand about this tool, it is purpose is kind of developer tool which run analysis on the common npm modules and verifies how analysis goes on then. This is not actual automated testing, but rather tool for capturing feedback. I maybe wrong about that, but I use it that way, trying figure out what's wrong. Again @DinoV maybe give clearer explanation.

Comment (mousetraps) [inline Nodejs/Tests/AnalysisTests/AnalysisFile.cs]: Ahh, I think I understand where you're coming from now...

It's not "automated testing" per se, but it is still important that we are testing the code that is in the product. Otherwise if we break something in the product related to this, we won't figure it out. The analysis limits are getting set from the following location.
https://github.com/Microsoft/nodejstools/blob/master/Nodejs/Tests/AnalysisDriver/AnalysisDriver.cs#L170

So we should provide an option to also run with medium analysis set (instead of the boolean flag we have right now to distinguish between low/high)

Comment (mousetraps) [inline Nodejs/Tests/AnalysisTests/AnalysisFile.cs]: Essentially we use it to make sure analysis continues performing as expected as we change things and as the packages themselves change.

Comment (DinoV) [inline Nodejs/Tests/AnalysisTests/AnalysisFile.cs]: Really the tool exists to do comparisons between different versions of the analyzer.  It certainly doesn't run anywhere in an automated way because you really need to get your baseline, make the changes, and then select what you want to compare it to.

For keeping the product code and test code in sync here what I'd do is move the file check into a common helper where both NTVS and AnalysisDriver can both call into it (e.g. JsAnalyzer).  Then I'd have them both manage the nested modules limit independently - with AnalysisDriver updated to take it from the command line, and NTVS taking it from user settings somewhere.

Comment (kant2002) [inline Nodejs/Tests/AnalysisTests/AnalysisFile.cs]: I try to find place where this could be placed, but did not find any public helper class which accessible from AnalysisDryver.

@DinoV could you show me 
1. in which public helper could I place this code 
2. or alternatively I could add InternalsVisibleTo("AnalysisDriver") to Analyisys assembly.

Comment (DinoV) [inline Nodejs/Tests/AnalysisTests/AnalysisFile.cs]: I would add it to AnalysisLimits, that seems like a reasonable spot for it to go

Comment (kant2002) [inline Nodejs/Tests/AnalysisTests/AnalysisFile.cs]: Reasonable place. Already done that. @DinoV and @mousetraps give last look, hopefully this is last comments :smile:

File: Nodejs/Product/Analysis/Analysis/AnalysisLimits.cs [hunk 18]
```diff
@@ -34,6 +34,44 @@ internal class AnalysisLimits {
...
+            // Level 0 - Large system. Some code should use properties of the object on level 4 here. 
+            // Level 1 - Subsystem - pass data from level 4
+            // Level 2 - Internal dependency for company - Wrap access to internal library and perform business logic. Do some work and pass data to level 1
+            // Level 3 - Internal library which wrap access to API.
+            // Level 4 - Http library.
+            // Level 5 - Promise Polyfill.
+            //
+            // All these examples are highly speculative and I specifically try to create such deep level. 
+            // If you develop on windows with such deep level you already close to your limit, your maximum is probably 10. 
+            // </remarks>
```
Comment (mousetraps) [inline Nodejs/Product/Analysis/Analysis/AnalysisLimits.cs]: nit: there's an end tag, but no opening tag here

**Comment by mousetraps** [comment]
Awesome, looks great! super nitpicky comment, but overall :+1:

File: Nodejs/Product/Analysis/Analysis/AnalysisLimits.cs (hunk 18 above)
Comment (kant2002) [inline Nodejs/Product/Analysis/Analysis/AnalysisLimits.cs]: Just remove it.
